ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommittingSpec.scala

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils
+
+import com.codahale.metrics.MetricRegistry
+import com.daml.ledger.participant.state.v1.{
+  SubmitterInfo,
+  ApplicationId,
+  CommandId,
+  TransactionMeta,
+}
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{DamlStateKey, DamlCommandDedupKey}
+import com.daml.lf.crypto
+import com.daml.lf.data.Time
+import com.daml.lf.data.Ref.{Party, Identifier}
+import com.daml.lf.data.ImmArray
+import com.daml.lf.transaction.SubmittedTransaction
+import com.daml.lf.transaction.test.TransactionBuilder
+import com.daml.lf.value.Value
+import com.daml.metrics.Metrics
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+import java.time.Instant
+
+class KeyValueCommitingSpec extends AnyWordSpec with Matchers {
+  private val metrics: Metrics = new Metrics(new MetricRegistry)
+  private val keyValueSubmission = new KeyValueSubmission(metrics)
+
+  private val alice = Party.assertFromString("Alice")
+  private val commandId = CommandId.assertFromString("cmdid")
+
+  private def toSubmission(tx: SubmittedTransaction) = {
+    val timestamp = Time.Timestamp.assertFromInstant(Instant.EPOCH)
+    val meta = TransactionMeta(
+      ledgerEffectiveTime = timestamp,
+      workflowId = None,
+      submissionTime = timestamp,
+      submissionSeed = crypto.Hash.hashPrivateKey(this.getClass.getName),
+      optUsedPackages = Some(Set.empty),
+      optNodeSeeds = None,
+      optByKeyNodes = None,
+    )
+    val submitterInfo = SubmitterInfo(
+      actAs = List(alice),
+      applicationId = ApplicationId.assertFromString("appid"),
+      commandId = commandId,
+      deduplicateUntil = Instant.EPOCH,
+    )
+    keyValueSubmission.transactionToSubmission(
+      submitterInfo = submitterInfo,
+      meta = meta,
+      tx = tx,
+    )
+  }
+
+  def getOutputs(builder: TransactionBuilder) =
+    KeyValueCommitting.submissionOutputs(toSubmission(builder.buildSubmitted()))
```

Comment:
Weirds me out a little that the method under test is hidden away in a helper. I worry it makes it hard to figure out what we're testing here.

Perhaps pull out `toSubmission(builder.buildSubmitted())` into a helper method but keep `submissionOutputs` in the test cases?

Reply:
e02c75be8f8ff772fc44938c76a7bc3e6ee31ef1